### PR TITLE
adding password recovery apis

### PR DESCRIPTION
**Pull request by GayanM**

### CODE REVIEW

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/pom.xml [hunk 1]
```diff
@@ -37,6 +37,16 @@
             <artifactId>commons-lang</artifactId>
         </dependency>
         <dependency>
+            <groupId>org.apache.cxf</groupId>
+            <artifactId>cxf-bundle</artifactId>
+            <version>2.7.13</version>
```
Comment (thanujalk) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/pom.xml]: move version t dependency management

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/pom.xml [hunk 2]
```diff
@@ -37,6 +37,16 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.jaxrs</groupId>
+            <artifactId>jackson-jaxrs-json-provider</artifactId>
+            <version>2.4.2</version>
```
Comment (thanujalk) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/pom.xml]: move version t dependency management

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/PasswordRecoverySecurityQuestionClient.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import javax.ws.rs.core.Response;
+
+public class PasswordRecoverySecurityQuestionClient {
+
+    StringBuilder builder = new StringBuilder();
+    /*String url = builder.append(IdentityManagementServiceUtil.getInstance().getServiceContextURL()).append
+            (IdentityManagementEndpointConstants.ServiceEndpoints.USER_INFORMATION_RECOVERY_SERVICE).toString()
+            .replaceAll("(?<!(http:|https:))//", "/")+"/account-recovery";*/
+
+    String url = "http://localhost:9763"+"/account-recovery";
```
Comment (thanujalk) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/PasswordRecoverySecurityQuestionClient.java]: remove +

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/PasswordRecoverySecurityQuestionClient.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    public ChallengeQuestionsResponse verifyUserChallengeAnswerAtOnce (VerifyAllAnswerRequest verifyAllAnswerRequest){
+        PasswordRecoverySecurityQuestion passwordRecoverySecurityQuestion = JAXRSClientFactory.create(url, PasswordRecoverySecurityQuestion.class,IdentityManagementServiceUtil.getInstance().getJSONProvider());
+        Response response = passwordRecoverySecurityQuestion.verifyUserChallengeAnswerAtOnce(verifyAllAnswerRequest);
+        ChallengeQuestionsResponse challengeQuestionsResponse = response.readEntity(ChallengeQuestionsResponse.class);
+        return challengeQuestionsResponse;
+    }
+
+}
```
Comment (thanujalk) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/PasswordRecoverySecurityQuestionClient.java]: format the code. length exceeds 120.

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/PasswordRecoverySecurityQuestionClient.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
+package org.wso2.carbon.identity.mgt.endpoint.serviceclient;
```
Comment (thanujalk) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/PasswordRecoverySecurityQuestionClient.java]: licence header

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/PasswordRecoverySecurityQuestionClient.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+import org.apache.cxf.jaxrs.client.JAXRSClientFactory;
+import org.wso2.carbon.identity.mgt.endpoint.IdentityManagementEndpointConstants;
+import org.wso2.carbon.identity.mgt.endpoint.IdentityManagementServiceUtil;
+import org.wso2.carbon.identity.mgt.endpoint.serviceclient.beans.*;
```
Comment (thanujalk) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/PasswordRecoverySecurityQuestionClient.java]: remove *

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/ChallengeQuestionResponse.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
+package org.wso2.carbon.identity.mgt.endpoint.serviceclient.beans;
```
Comment (thanujalk) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/ChallengeQuestionResponse.java]: licence header

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/ChallengeQuestionsResponse.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
+package org.wso2.carbon.identity.mgt.endpoint.serviceclient.beans;
```
Comment (thanujalk) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/ChallengeQuestionsResponse.java]: licence header

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/UserPassword.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+    public User getUser() {
+        return user;
+    }
+
+    public void setUser(User user) {
+        this.user = user;
+    }
+
+    public String getPassword() {
```
Comment (thanujalk) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/UserPassword.java]: can we use char array for password?

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/PasswordRecoverySecurityQuestionClient.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+    StringBuilder builder = new StringBuilder();
+    /*String url = builder.append(IdentityManagementServiceUtil.getInstance().getServiceContextURL()).append
+            (IdentityManagementEndpointConstants.ServiceEndpoints.USER_INFORMATION_RECOVERY_SERVICE).toString()
+            .replaceAll("(?<!(http:|https:))//", "/")+"/account-recovery";*/
+
+    String url = "http://localhost:9763"+"/account-recovery";
+
+    public ChallengeQuestionResponse initiateUserChallengeQuestion (User user){
+        PasswordRecoverySecurityQuestion passwordRecoverySecurityQuestion = JAXRSClientFactory.create(url, PasswordRecoverySecurityQuestion.class, IdentityManagementServiceUtil.getInstance().getJSONProvider());
```
Comment (omindu) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/PasswordRecoverySecurityQuestionClient.java]: Exceeding 120 column width.

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/ChallengeQuestionResponse.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+import org.wso2.carbon.identity.mgt.endpoint.serviceclient.model.ChallengeQuestion;
+
+import javax.xml.bind.annotation.*;
```
Comment (omindu) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/ChallengeQuestionResponse.java]: Avoid using * imports

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/ChallengeQuestionResponse.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public String getStatus() {
+        return status;
+    }
+
+    public void setStatus(String status) {
+        this.status = status;
+    }
+}
```
Comment (omindu) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/ChallengeQuestionResponse.java]: End file with newline.

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/PasswordRecoverySecurityQuestionClient.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+import org.apache.cxf.jaxrs.client.JAXRSClientFactory;
+import org.wso2.carbon.identity.mgt.endpoint.IdentityManagementEndpointConstants;
+import org.wso2.carbon.identity.mgt.endpoint.IdentityManagementServiceUtil;
+import org.wso2.carbon.identity.mgt.endpoint.serviceclient.beans.*;
+import org.wso2.carbon.identity.mgt.endpoint.serviceclient.client.proxy.api.PasswordRecoverySecurityQuestion;
+
+import javax.ws.rs.core.Response;
+
+public class PasswordRecoverySecurityQuestionClient {
```
Comment (omindu) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/PasswordRecoverySecurityQuestionClient.java]: Add a class level comment.

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/ChallengeQuestionResponse.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
+package org.wso2.carbon.identity.mgt.endpoint.serviceclient.beans;
+
+import org.wso2.carbon.identity.mgt.endpoint.serviceclient.model.ChallengeQuestion;
+
+import javax.xml.bind.annotation.*;
+
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlRootElement(name = "")
+public class ChallengeQuestionResponse {
```
Comment (omindu) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/ChallengeQuestionResponse.java]: Add a class level comment.

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/UserPassword.java [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.mgt.endpoint.serviceclient.beans;
+
+import javax.xml.bind.annotation.*;
```
Comment (omindu) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/UserPassword.java]: Avoid using * imports.

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/VerifyAllAnswerRequest.java [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.mgt.endpoint.serviceclient.beans;
+
+import org.wso2.carbon.identity.mgt.endpoint.serviceclient.model.UserChallengeAnswer;
+
+import javax.xml.bind.annotation.*;
```
Comment (omindu) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/serviceclient/beans/VerifyAllAnswerRequest.java]: Avoid using * imports.

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/IdentityManagementServiceUtil.java [hunk 17]
```diff
@@ -41,6 +44,8 @@
 public class IdentityManagementServiceUtil {
 
     private static IdentityManagementServiceUtil instance = new IdentityManagementServiceUtil();
+    private JSONProvider jsonProvider = new JSONProvider ();
```
Comment (thanujalk) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/IdentityManagementServiceUtil.java]: unnecessary space in 'new JSONProvider ()'

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/IdentityManagementServiceUtil.java [hunk 18]
```diff
@@ -142,6 +152,10 @@ public void authenticate(ServiceClient client) {
         option.setManageSession(true);
     }
 
+    public List getJSONProvider (){
```
Comment (thanujalk) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/IdentityManagementServiceUtil.java]: unnecessary space after method name

**Pull request merged**